pkg/envoy/ads/response.go

```diff
@@ -18,74 +19,80 @@ const (
 )
 
 // Wrapper to create and send a discovery response to an envoy server
-func (s *Server) sendTypeResponse(tURI envoy.TypeURI,
-	proxy *envoy.Proxy, server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer,
-	req *xds_discovery.DiscoveryRequest, cfg configurator.Configurator) error {
+func (s *Server) sendTypeResponse(typeURI envoy.TypeURI, proxy *envoy.Proxy, server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer, req *xds_discovery.DiscoveryRequest, cfg configurator.Configurator) error {
 	// Tracks the success of this TypeURI response operation; accounts also for receipt on envoy server side
-	success := false
-	xdsShortName := envoy.XDSShortURINames[tURI]
-	defer xdsPathTimeTrack(time.Now(), log.Debug(), xdsShortName, proxy.GetCertificateSerialNumber().String(), &success)
+	startedAt := time.Now()
+	log.Trace().Msgf("[%s] Creating response for proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
 
-	log.Trace().Msgf("[%s] Creating response for proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-
-	discoveryResponse, err := s.newAggregatedDiscoveryResponse(proxy, req, cfg)
-	if err != nil {
-		log.Error().Err(err).Msgf("[%s] Failed to create response for proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+	if discoveryResponse, err := s.newAggregatedDiscoveryResponse(proxy, req, cfg); err != nil {
+		log.Error().Err(err).Msgf("[%s] Failed to create response for proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+		xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, false)
 		return err
-	}
-
-	if err := (*server).Send(discoveryResponse); err != nil {
-		log.Error().Err(err).Msgf("[%s] Error sending to proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+	} else if err := (*server).Send(discoveryResponse); err != nil {
+		log.Error().Err(err).Msgf("[%s] Error sending to proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+		xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, false)
 		return err
 	}
 
-	success = true // read by deferred function
+	xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, true)
```

Comment:
To me, having to call this function at every exit point is not an improvement, it makes the code more prone to error as anyone changing this code could forget to call it on a new return. the defer was at least ensuring the call was made no matter the exit point.

I don't know if the tradeoff is readability vs function, but I'd not choose the former one here.

Reply:
I agree - this is a concern.  The Go `defer` is great in these cases, although never free.
My main concern here was with the `success` variable.  I was initially surprised by the usage of a pointer to a bool, which seems unusual and it took me awhile to understand "what is going on here".
And every time I think "what is going on here" - I know that I won't be the last one to ask that (I myself will probably wonder the same in a week when I read the same piece of code).

So I decided that it would add clarity to remove the pointer to a bool, and instead copy/paste the function invocation at all returns.

Reply:
>My main concern here was with the success variable

you got me there. I didn't like that either.